File: binary_search/binary_search.go
package binarySearch

import "errors"

// ErrNotFound for no element in the sorted array
var ErrNotFound = errors.New("could not find the element in the array")

// BinarySearch searches for an element in a sorted array and returns its index
func BinarySearch(arr []int, item int) (int, error) {
	low := 0
	high := len(arr) - 1

	for low <= high {
		mid := (low + high) / 2
		if arr[mid] == item {
			return mid, nil
		} else if arr[mid] < item {
			low = mid + 1
		} else {
			high = mid - 1
		}
	}

	return -1, ErrNotFound
}

File: binary_search/binary_search.py
def binary_search(arr, item):
    """Binary Search assumes that the input array is sorted."""
    low = 0
    high = len(arr) - 1

    while low <= high:
        mid = (low + high) // 2
        guess = arr[mid]

        if guess == item:
            return guess
        elif guess < item:
            low = mid + 1
        else:
            high = mid - 1

    return None


def run_test():
    assert binary_search([1, 2, 3, 4, 5, 6], 6) == 6
    assert binary_search([1, 2, 3, 4, 5, 6], 0) is None
    print("tests passed...")


if __name__ == "__main__":
    run_test()
